### PR TITLE
storage/engine: tweak RocksDB options

### PR DESCRIPTION
**Pull request by petermattis**
Together these tweaks reduce the time for `block_writer` to insert 1m
rows from 250s to 225s (-10%).

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7532)

### CODE REVIEW

**Comment by petermattis** [comment]
These are pulled from my `pmattis/rocksdb-tuning` branch.

File: storage/engine/rocksdb/db.cc [hunk 1]
```diff
@@ -1353,15 +1353,9 @@ DBStatus DBOpen(DBEngine **db, DBSlice dir, DBOptions db_opts) {
...
-  cf_options.OptimizeLevelStyleCompaction(db_opts.memtable_budget);
-  // OptimizeLevelStyleCompaction sets no-compression for L0 and
-  // L1. Current benchmarks and tests show no benefit to doing this.
-  for (int i = 0; i < cf_options.compression_per_level.size(); i++) {
-    cf_options.compression_per_level[i] = rocksdb::kSnappyCompression;
-  }
-
-  rocksdb::Options options(rocksdb::DBOptions(), cf_options);
+  rocksdb::Options options(rocksdb::GetOptions(db_opts.memtable_budget));
+  options.IncreaseParallelism(db_opts.num_cpu);
```
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: Any commentary here?

Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: Which line? The `IncreaseParallelism` call allows multiple background compactions to be performed concurrently and ties into the `max_subcompactions` setting below which will use multiple threads to speed up a single compaction.

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/engine/rocksdb/db.cc, line 1368 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqCFCGnnwv4n9bMar:-KLSqCFDJJMIoi430i7p:1589016654) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc97dc47e56b2ede862902cf49ef267bdafea0f/storage/engine/rocksdb/db.cc#L1368)):*
> ```cpp
>   options.table_factory.reset(rocksdb::NewBlockBasedTableFactory(table_options));
> 
>   // File size settings: TODO(marc): investigate and determine better long-term settings:
> ```

this TODO is sitting on the wrong thing now

---

*[storage/engine/rocksdb/db.cc, line 1358 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqL89BJaqar6c626w:-KLSqL89BJaqar6c626x:-1709433982) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2868189cbfa6813924461da6acf08f5c4b9f90c/storage/engine/rocksdb/db.cc#L1358)):*
> ```cpp
>   rocksdb::Options options(rocksdb::GetOptions(db_opts.memtable_budget));
>   options.IncreaseParallelism(db_opts.num_cpu);
>   options.max_subcompactions = std::max<int>(db_opts.num_cpu / 2, 1);
> ```

science dog is surprised you needed to specify the type parameter

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-:-KLSqQvWT6NUj98qRigg:1417088721)*

File: storage/engine/rocksdb/db.cc (hunk 1 above)
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: Mostly the second but generally all of them - I need a comment to explain why `(num_cpu / 2, 1)` was chosen and I think everyone else who's not in the RocksDB internals will appreciate it too. Super useful to know what your reasoning is when you turn random knobs.

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532)*

**Comment by petermattis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/engine/rocksdb/db.cc, line 1368 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqCFCGnnwv4n9bMar:-KLSwMwv8A3JWnD8pvQn:1267029685) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc97dc47e56b2ede862902cf49ef267bdafea0f/storage/engine/rocksdb/db.cc#L1368)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this TODO is sitting on the wrong thing now
</details>
Removed as I've now done some tuning here.

---

*[storage/engine/rocksdb/db.cc, line 1357 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLScr6ijb860OtyGZ7Z-r3-1357:-KLSvkmbrIveaCAlxt44:1626901177) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2868189cbfa6813924461da6acf08f5c4b9f90c/storage/engine/rocksdb/db.cc#L1357)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Mostly the second but generally all of them - I need a comment to explain why `(num_cpu / 2, 1)` was chosen and I think everyone else who's not in the RocksDB internals will appreciate it too. Super useful to know what your reasoning is when you turn random knobs.
</details>
I added some commentary in various places.

---

*[storage/engine/rocksdb/db.cc, line 1358 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqL89BJaqar6c626w:-KLSvpEfI_AZkISh3yTU:1782943938) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2868189cbfa6813924461da6acf08f5c4b9f90c/storage/engine/rocksdb/db.cc#L1358)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> science dog is surprised you needed to specify the type parameter
</details>
It wasn't. Removed.

---

*[storage/engine/rocksdb/db.cc, line 1388 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSvs4qQZsylhCPsfx8:-KLSvs4qQZsylhCPsfx9:-469213201) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b8c30935fccd39434ccf4f7ee0976aed6f9f744/storage/engine/rocksdb/db.cc#L1388)):*
Note that I added tuning of this field which was massively misconfigured. With the current settings we'll compute the value here as 32 MB while we previously had it configured as 512 MB.

> We usually recommend making L0->L1 as fast as possible by making the size of level 0 similar to size of level 1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532)*

**Comment by tschottdorf** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-:-KLT9amhbSNQfkbjoWwm:1417088721)*

**Comment by cuongdo** [comment]
:lgtm:

Unfortunately, this doesn't improve performance for the other K/V slowdown (#7280), but these seem like good changes nonetheless based on RocksDB tuning guidelines.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-:-KLTyjHmRBJNHiAXblYO:928237250)*

**Pull request merged**